src/lib.rs

```diff
@@ -294,6 +294,29 @@ impl<T, E> Pool<T, E> {
         obj.state = ObjectState::Ready;
         Ok(obj)
     }
+    /// Add existing value to the pool. Returns whether or not the value was successfully added.
+    pub fn add(&self, value: T) -> bool {
+        let max_size = self.inner.config.max_size;
+        let mut current_size;
+        while {
+            // Get current size and check if max
+            current_size = self.inner.size.load(Ordering::Relaxed);
+            current_size < max_size
+        } {
+            // Try to increment size and push value
+            let old_size =
+                self.inner
+                    .size
+                    .compare_and_swap(current_size, current_size + 1, Ordering::Relaxed);
```

Comment:
While this makes sure that no two concurrent calls to `add` add more than `max_size` objects to the pool it is still possible to have a race between `add` and `get_timeout` which causes more objects being pushed to the queue than there are actual slots: https://github.com/bikeshedder/deadpool/blob/master/src/lib.rs#L280 - If there are `max_size` tasks currently in the critical section, but `size.fetch_add(1, Ordering::Relaxed)` has not been called yet you can end up adding an extra object to the queue and exceeding the `max_size` of the pool.

The current version of deadpool (0.4) relies entirely on the semaphore to ensure that there are no more than `max_size` tasks in that critical section. If the queue is empty it's safe to assume that `size` has not reached `max_size`, yet, and a new object can be created.

If it is really a viable use case to add existing object to a pool during runtime I'd change the `get_timeout` method to call `add` internally which uses the atomics to ensure that no more than `max_size` objects are ever added to the pool.